spec/rspec/rails/matchers/have_reported_error_spec.rb

```diff
@@ -0,0 +1,213 @@
+      }.to raise_error(RuntimeError, /Using the `have_reported_error` matcher with a `nil` error is probably unintentional/)
+    end
+
+    it "fails when no errors are reported" do
+      expect {
+        expect { "no error" }.to have_reported_error
+      }.to fail_with(/Expected the block to report an error, but none was reported./)
+    end
+
+    it "fails when no errors are reported with explicit nil argument" do
+      expect {
+        expect { "no error" }.to have_reported_error
+      }.to fail_with(/Expected the block to report an error, but none was reported./)
+    end
+
+    it "passes when negated and no errors are reported" do
+      expect { "no error" }.not_to have_reported_error
+    end
+  end
+
+  context "constrained to a specific error class" do
+    it "passes when an error with the correct class is reported" do
+      expect { Rails.error.report(TestError.new("test error")) }.to have_reported_error(TestError)
+    end
+
+    it "fails when an error with the wrong class is reported" do
+      expect {
+        expect {
+          Rails.error.report(AnotherTestError.new("wrong error"))
+        }.to have_reported_error(TestError)
+      }.to fail_with(/Expected error to be an instance of TestError, but got: AnotherTestError/)
+    end
+  end
+
+  context "constrained to a matching error (class and message)" do
+    it "passes with an error that matches exactly" do
+      expect {
+        Rails.error.report(TestError.new("exact message"))
+      }.to have_reported_error(TestError, "exact message")
+    end
+
+    it "passes any error of the same class if no message is specified" do
+      expect {
+        Rails.error.report(TestError.new("any message"))
+      }.to have_reported_error(TestError)
+    end
+
+    it "fails when the error has different message to the expected" do
+      expect {
+        expect {
+          Rails.error.report(TestError.new("actual message"))
+        }.to have_reported_error(TestError, "expected message")
+      }.to fail_with(/Expected error to be an instance of TestError with message 'expected message', but got: TestError/)
+    end
+  end
+
+  context "constrained by regex pattern matching" do
+    it "passes when an error message matches the pattern" do
+      expect {
+        Rails.error.report(StandardError.new("error with pattern"))
+      }.to have_reported_error(StandardError, /with pattern/)
+    end
+
+    it "fails when no error messages match the pattern" do
+      expect {
+        expect {
+          Rails.error.report(StandardError.new("error without match"))
+        }.to have_reported_error(StandardError, /different pattern/)
+      }.to fail_with(/Expected error to be an instance of StandardError with message/)
+    end
+  end
+
+  describe "#failure_message" do
+    it "provides details about mismatched attributes" do
+      expect {
+        expect {
+          Rails.error.report(StandardError.new("test"), context: { user_id: 123, context: "actual" })
```

Comment:
Nested `context` key is confusing here. Do you mind to change the inner one to something else?

Reply:
Uh, resolved.